registration/include/pcl/registration/impl/correspondence_estimation.hpp

```diff
@@ -126,45 +127,64 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::determineCorresponde
 
   std::vector<int> index(1);
   std::vector<float> distance(1);
-  pcl::Correspondence corr;
-  unsigned int nr_valid_correspondences = 0;
+  std::atomic_uint nr_valid_correspondences(0);
 
   // Check if the template types are the same. If true, avoid a copy.
   // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
   // macro!
   if (isSamePointType<PointSource, PointTarget>()) {
     // Iterate over the input set of source indices
-    for (const auto& idx : (*indices_)) {
+    #pragma omp parallel for \
+      default(none) \
+      shared(tree_, indices_, max_dist_sqr, correspondences, nr_valid_correspondences) \
+      firstprivate(index, distance) \
+      num_threads(num_threads_)
+    for (int i = 0; i < static_cast<int>(indices_->size()); i++) {
+      const auto& idx = (*indices_)[i];
       tree_->nearestKSearch((*input_)[idx], 1, index, distance);
       if (distance[0] > max_dist_sqr)
         continue;
 
+      pcl::Correspondence corr;
       corr.index_query = idx;
       corr.index_match = index[0];
       corr.distance = distance[0];
       correspondences[nr_valid_correspondences++] = corr;
     }
   }
   else {
-    PointTarget pt;
-
     // Iterate over the input set of source indices
-    for (const auto& idx : (*indices_)) {
+    #pragma omp parallel for \
+      default(none) \
+      shared(tree_, indices_, max_dist_sqr, correspondences, nr_valid_correspondences) \
+      firstprivate(index, distance) \
+      num_threads(num_threads_)
+    for (int i = 0; i < static_cast<int>(indices_->size()); i++ ) {
+      const auto& idx = (*indices_)[i];
+
       // Copy the source data to a target PointTarget format so we can search in the
       // tree
+      PointTarget pt;
       copyPoint((*input_)[idx], pt);
 
       tree_->nearestKSearch(pt, 1, index, distance);
       if (distance[0] > max_dist_sqr)
         continue;
 
+      pcl::Correspondence corr;
       corr.index_query = idx;
       corr.index_match = index[0];
       corr.distance = distance[0];
       correspondences[nr_valid_correspondences++] = corr;
     }
   }
   correspondences.resize(nr_valid_correspondences);
+
+  if(num_threads_ != 1) {
+    // Make correspondences ordered
+    std::sort(correspondences.begin(), correspondences.end(), [](const auto& lhs, const auto& rhs) { return lhs.index_query < rhs.index_query; });
```

Comment:
Good suggestion. I compared the following strategies:
1. Use atomic and sort (the first commit)
2. Separate correspondences per-thread and merge them after the loop (the last commit)
3. Do binary search and insert in critical section
4. Use OMP user-defined reduction

Here is a short summary of the benchmark results, and I think the 2nd strategy is the best one.
1. Sort gets slower as the number of threads increases
2. Almost constant merge time and small overhead
3. It was the slowest as expected
4. No post process time, but slightly larger overhead

Regarding the part of ```inplace_merge```, although it would be better to merge results in a divide-and-conquer way, I think it makes the code much complex. Since we can expect ```num_threads``` is not so large, I consider the current form (merging results sequentially) is sufficient. 

Benchmark results:
```
# 69088pts vs 69792pts
# CorrespondenceEstimation::determineCorrespondences()

# without omp (baseline)
num_threads:total time
1:169.906[msec]
2:177.868[msec]
3:172.066[msec]
4:169.009[msec]
5:169.16[msec]

# 1. atomic and sort
num_threads: for loop, post process(sort), total time
1:169.316[msec] 5e-05[msec] 169.327[msec]
2:98.9463[msec] 1.64847[msec] 100.606[msec]
3:108.082[msec] 2.1331[msec] 110.225[msec]
4:66.546[msec] 2.39187[msec] 68.9498[msec]
5:62.1945[msec] 2.65166[msec] 64.8581[msec]
6:52.2621[msec] 3.33361[msec] 55.6112[msec]
7:60.9466[msec] 2.83809[msec] 63.7972[msec]
8:45.5633[msec] 2.79004[msec] 48.3653[msec]
9:48.2131[msec] 2.88753[msec] 51.1126[msec]

# 2. per-thread correspondences and inplace_merge
num_threads: for loop, post process(merge), total time
1:167.489[msec] 4.8e-05[msec] 167.502[msec]
2:93.6329[msec] 0.490607[msec] 94.138[msec]
3:100.873[msec] 0.604739[msec] 101.493[msec]
4:65.5067[msec] 0.402887[msec] 65.9252[msec]
5:61.0197[msec] 0.674917[msec] 61.7172[msec]
6:60.6732[msec] 0.529535[msec] 61.2234[msec]
7:50.8765[msec] 0.557357[msec] 51.449[msec]
8:61.3818[msec] 0.605387[msec] 62.0017[msec]
9:48.0522[msec] 0.634407[msec] 48.7004[msec]

# 3. insert in critical section
num_threads: total time
1:3174.6[msec]
2:4121.34[msec]
3:4462.44[msec]
4:4753.24[msec]
5:4854.75[msec]

# 4. OMP user-defined reduction
num_threads: for loop, post process, total time
1:174.546[msec] 4.6e-05[msec] 174.638[msec]
2:94.445[msec] 5.2e-05[msec] 94.4573[msec]
3:102.331[msec] 7.6e-05[msec] 102.343[msec]
4:68.7443[msec] 0.000117[msec] 68.8659[msec]
5:68.1277[msec] 4e-05[msec] 68.1397[msec]
6:53.522[msec] 4.4e-05[msec] 53.5339[msec]
7:62.1336[msec] 0.000172[msec] 62.1449[msec]
8:57.7286[msec] 0.000171[msec] 57.7404[msec]
9:58.9015[msec] 0.000131[msec] 58.915[msec]
```

Code of the 4th strategy (omp reduction) was something like the following one:
```cpp
  #pragma omp declare reduction \
    (merge: pcl::Correspondences : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()) )
  pcl::Correspondences corrs;

  #pragma omp parallel for \
    default(none) \
    shared(tree_, indices_, max_dist_sqr) \
    firstprivate(index, distance) \
    reduction(merge : corrs) \
    num_threads(num_threads_)
  for (int i = 0; i < static_cast<int>(indices_->size()); i++) {
    const auto& idx = (*indices_)[i];
    tree_->nearestKSearch((*input_)[idx], 1, index, distance);
    if (distance[0] > max_dist_sqr)
      continue;

    pcl::Correspondence corr;
    corr.index_query = idx;
    corr.index_match = index[0];
    corr.distance = distance[0];

    corrs.push_back(corr);
  }

  correspondences.swap(corrs);

```

Reply:
Nice experiment @koide3 ! :heart:
Thanks a lot for the follow-through. Leaving this un-resolved so others can see this too :smile: